File: 30_substring_with_concatenation_of_all_words/main.py
# -*- coding: utf-8 -*-


class Solution(object):

    def find_substring(self, s, words):
        res = []
        if len(words) == 0 or (len(words) > 0 and words[0] == ''):
            return res
        words_count = {}
        for word in words:
            if word not in words_count:
                words_count[word] = 0
            words_count[word] += 1

        s_len, e_len = len(words) * len(words[0]), len(words[0])
        for i in range(0, len(s) - s_len + 1):
            if s[i:i + e_len] not in words:
                continue
            ss = s[i:i + s_len]
            tmp_count, skip = {}, False
            for j in range(0, s_len, e_len):
                word = ss[j:j + e_len]
                if word not in words:
                    skip = True
                    break
                if word not in tmp_count:
                    tmp_count[word] = 0
                tmp_count[word] += 1
            if skip:
                continue
            if tmp_count == words_count:
                res.append(i)
        return res


if __name__ == '__main__':
    s = Solution()
    print(s.find_substring('barfoothefoobarman', ['foo', 'bar']))

File: 30_substring_with_concatenation_of_all_words/go/main.go
package main

import (
	"fmt"
	"reflect"
)

func findSubstring(s string, words []string) []int {
	res := make([]int, 0, len(s) / 2)
	if len(words) == 0 || (len(words) > 0 && len(words[0]) == 0) {
		return res
	}
	s_len, e_len := len(words) * len(words[0]), len(words[0])
	words_count := make(map[string]int)
	for _, word := range words {
		words_count[word]++
	}

	for i := 0; i <= len(s) - s_len; i++ {
		if words_count[s[i:i + e_len]] == 0 {
			continue
		}
		ss, tmp_count, skip := s[i:i + s_len], make(map[string]int), false
		for j := 0; j < s_len; j += e_len {
			word := ss[j:j + e_len]
			if words_count[word] == 0 {
				skip = true
				break
			}
			tmp_count[word]++
		}
		if skip {
			continue
		}
		if reflect.DeepEqual(tmp_count, words_count) {
			res = append(res, i)
		}
	}

	return res
}

func main() {
	fmt.Println(findSubstring("barfoothefoobarman", []string{"foo", "bar"}))
}
